tutorials/finewebpdfs/download.py

```diff
@@ -0,0 +1,269 @@
+"""
+Tutorial: Download PDFs from Common Crawl using the FinePDFs index.
+
+Pipeline steps:
+- URL generation from the HuggingFaceFW/finepdfs dataset. Each entry contains a WARC `file_path` and an `offset`.
+  We convert `s3://commoncrawl` to `https://data.commoncrawl.org` and embed the offset as a URL fragment
+  (e.g., https://data.commoncrawl.org/crawl-data/CC-MAIN-.../segments/.../warc/....warc.gz#offset=123456789).
+- Download the WARC files with CommonCrawlWARCStartPosDownloader (wget with --start-pos).
+- Save PDF bytes at the given offset as `original_file_name_offset.pdf` into a target directory.
+"""
+
+import argparse
+import json
+import os
+import pickle
+import subprocess
+from dataclasses import dataclass
+from urllib.parse import parse_qs, urlparse
+
+from loguru import logger
+from warcio.archiveiterator import ArchiveIterator
+
+from nemo_curator.backends.xenna import XennaExecutor
+from nemo_curator.core.client import RayClient
+from nemo_curator.pipeline import Pipeline
+from nemo_curator.stages.base import ProcessingStage
+from nemo_curator.stages.resources import Resources
+from nemo_curator.stages.text.download.base.download import DocumentDownloader, DocumentDownloadStage
+from nemo_curator.stages.text.download.base.url_generation import URLGenerationStage, URLGenerator
+from nemo_curator.tasks import FileGroupTask
+
+
+class FinePDFsURLGenerator(URLGenerator):
+    """Generate Common Crawl HTTPS WARC URLs with embedded offset from the FinePDFs dataset.
+
+    Each generated URL includes a URL fragment `#offset=...` to propagate the byte offset
+    through subsequent stages without changing stage interfaces.
+    """
+
+    def __init__(self, output_dir: str, streaming: bool = True) -> None:
+        self.output_dir = output_dir
+        self.streaming = streaming
+
+    def generate_urls(self) -> list[str]:
+        with open("/raid/praateekm/NeMo-Curator/finepdfs.jsonl") as f:
+            dataset = [json.loads(line) for line in f]
+
+        urls: list[str] = []
+        replace_prefix = ("s3://commoncrawl", "https://data.commoncrawl.org")
+        for row in dataset:  # type: ignore[reportTypedDictNotRequiredAccess]
+            file_path: str = row["file_path"]
+            offset_val: int = int(row["offset"])  # ensure int for naming
+
+            # Skip parquet index files - only process actual WARC files
+            if file_path.endswith(".parquet") or "/cc-index/table/" in file_path:
+                continue
```

Comment:
I'm not sure why bunch of files are index files. So for now I've skipped them. A more thorough approach will include them too